tests/framework/scheduler.py

```diff
@@ -0,0 +1,329 @@
+# Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+# SPDX-License-Identifier: Apache-2.0
+"""Pytest plugin that schedules tests to run concurrently.
+
+This plugin adds a new command line option (`--concurrency`), allowing the
+user to choose the maximum number of worker processes that can run tests
+concurrently.
+
+Tests are split into batches, each batch being assigned a maximum concurrency
+level. For instance, all performance tests will run sequentially
+(i.e. concurrency=1), since they rely on the availability of the full host
+resources, in order to make accurate measurements. Additionally, other tests
+may be restricted to running sequentially, if they are per se
+concurrency-unsafe. See `PytestScheduler.pytest_runtestloop()`.
+
+Scheduling is achieved by overriding the pytest run loop (i.e.
+`pytest_runtestloop()`), and splitting the test session item list across
+multiple `fork()`ed worker processes. Since no user code is run before
+`pytest_runtestloop()`, each worker becomes a pytest session itself.
+Reporting is disabled for worker process, each worker sending its results
+back to the main / server process, via an IPC pipe, for aggregation.
+"""
+
+import multiprocessing as mp
+import os
+import re
+import sys
+from random import random
+from select import select
+from time import sleep
+
+from . import mpsing  # pylint: disable=relative-beyond-top-level
+
+
+class PytestScheduler(mpsing.MultiprocessSingleton):
+    """A pretty custom test execution scheduler."""
+
+    def __init__(self):
+        """Initialize the scheduler.
+
+        Not to be called directly, since this is a singleton. Use
+        `PytestScheduler.instance()` to get the scheduler object.
+        """
+        super().__init__()
+        self._mpsings = [self]
+        self.session = None
+
+    def register_mpsing(self, mpsing_inst):
```

Comment:
nit: we can rename this to register_mp_singleton or register_mp_sing as it is easier to know what the method does when you have a "space" between the two words.

Reply:
Done. Naming is hard.